### PR TITLE
Prevent creating folders with trailing space & check for invalid characters

### PR DESCRIPTION
**Pull request by netwavebe**
Folders should not have leading or trailing whitespace characters.

Creation succeeds but it's impossible to upload anything to these folders. They also can't be removed easily (at least not on Windows).

### CODE REVIEW

**Comment by Skrypt** [comment]
Slugify them instead

**Comment by netwavebe** [comment]
Mmm, interesting idea!

Pros: always valid urls for media.
Cons: can't have spaces between words nor use PascalCasing.

"ProductImages" would become "productimages"
"Product Images" would become "product-images"

Still, I like this approach. Any reason not to slugify folders @deanmarcussen ?

**Comment by Skrypt** [comment]
Casing could be also slugified then.

**Comment by netwavebe** [comment]
The same could be said for filenames.

**Comment by deanmarcussen** [comment]
> Slugify them instead

I'm ok with just trimming here.

Slugify could be a feature.

But then someone will want pascal. and someone else will want camel...

Slugify would also strip diacritics in its current implement I think - which you made me fix @Skrypt :)

I do think this should be at service level, not in the controller though.

Blob storage accepts trailing spaces fine (but I've not checked the media app works with them. Imagine it does)

**Comment by deanmarcussen** [comment]
> I do think this should be at service level, not in the controller though.

`Combine` I meant. Or the `NormalizePath` extension.

**Comment by netwavebe** [comment]
@deanmarcussen Thanks for the feedback.

I found another issue when using a slash in the new folder name, like 'clients/suppliers'. This (obviously) creates a folder "clients" with another folder "suppliers" inside.   When naming the new folder 'clients / suppliers' I'm still getting folders with trailing/leading spaces... 😞

Don't think this is something I can fix in `Combine` or `NormalizePath`.

**Comment by netwavebe** [comment]
@deanmarcussen Please check the last commit when you have time.

Had to check for `\` and `/` in the controller as `_mediaFileStore.Combine(path, name);` should be able to combine `aaa/bbb` and `ccc/ddd`. However, in this particular case the second part should not contain nested folders.

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs [hunk 1]
```diff
@@ -377,6 +377,8 @@ public async Task<IActionResult> MoveMediaList(string[] mediaNames, string sourc
                 path = "";
             }
 
+            name = name.Replace('\\', '-').Replace('/', '-'); // Don't create nested folders
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs:380]: I'm not sure I agree with this change.

I would have thought it creating nested folders was a logical conclusion?

Seems like something we could discuss endlessly :)

**Comment by Skrypt** [comment]
@deanmarcussen Why would we want to keep filenames without slugifying them?
What does Wordpress do in his media gallery? Do they slugify filenames and paths?

Azure Blog Storage allows to use trailing spaces on folder/filenames ; means in url's they will be html encoded to `%20`. As far as I remember, those are not SEO friendly. Maybe we could have a global media gallery setting to slugify all paths/files then. This way we could keep natural file/folder names or not. But let's see what other CMS's does too.

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs (hunk 1 above)
Comment (netwavebe) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs:380]: It does create nested folders on Windows. For example, adding a folder named ´clients/suppliers' works fine (I get a folder 'clients' with another nested folder 'suppliers' inside).

Problem is when I add a folder named 'clients / supplier'. This adds a folder 'clients ' (with a trailing space!) with another nested folder ' suppliers' (with a leading space!) inside.

Folders with leading spaces are no problem, but trailing spaces are! You can't upload anything to it and you can't delete it (not in media library, not in file explorer). You would need a funky command like `rd "\\?\C:\path\media\clients "` to get rid of it.

I could split on `/`, trim each part and join again. Then I thought it probably wasn't the users intention to create nested folders, hence the change above.

**Comment by Skrypt** [comment]
Ok let's not get carried away and fix the trailing space issue for now. Forget the slugify thing.

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs (hunk 1 above)
Comment (netwavebe) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs:380]: @deanmarcussen FYI: even in the case when valid nested folders are created (valid = without trailing spaces) the media library doesn't handle this correctly. It shows the deepest folder as a direct descendant of the parent. We should prevent creating nested folders imho.

Comment (netwavebe) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs]: @deanmarcussen You're right!

**Comment by netwavebe** [comment]
`Create Folder` now shows a message if the folder name contains invalid characters:

createfolder

I agree this solution is much better.

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs [hunk 2]
```diff
@@ -377,6 +377,12 @@ public async Task<IActionResult> MoveMediaList(string[] mediaNames, string sourc
                 path = "";
             }
 
+            var invalidFolderNameCharacters = new char[] { '\\', '/' };
+            if (invalidFolderNameCharacters.Any(invalidChar => name.Contains(invalidChar)))
+            {
+                return StatusCode(StatusCodes.Status403Forbidden, S["Cannot create folder because the folder name contains invalid characters"]);
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs:383]: Looks great!

Call me Mr Picky but isn't model state errors `BadRequest` ?

https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.statuscodes.status400badrequest?view=aspnetcore-5.0

Comment (netwavebe) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs]: @deanmarcussen Yes, it is, I changed it.

**Review by deanmarcussen — APPROVED** [review]

**Comment by deanmarcussen** [comment]
@Skrypt @netwavebe I still like the idea of slugifying, just as a feature if anyone wants to do it.

(I would use it!)

**Comment by netwavebe** [comment]
@deanmarcussen @Skrypt I would use it too when starting on a new website. Might do this.

**Pull request merged**